File: apps/tabtin-ios/Tabtin/Features/Conversation/MessageBubble/StreamingMarkdownContinuityPolicy.swift
import Foundation

/// 流式才切稳定区 / 尾巴。收束不重新切分全文：
/// 若仍是流式期冻住的前缀，只把尾巴转成 Markdown；对不上再整段 parse。
enum StreamingMarkdownContinuityPolicy {
    enum TailRenderer: Equatable {
        case plainText
        case markdown
    }

    struct Layout: Equatable {
        let stable: String
        let tail: String
        let tailRenderer: TailRenderer

        var hasStable: Bool { !stable.isEmpty }
        var stableIdentity: String { stable }
    }

    static func layout(
        content: String,
        isStreaming: Bool,
        lastStreamingStable: String = ""
    ) -> Layout {
        if isStreaming {
            let parts = StreamingMarkdownSplitter.split(content)
            return Layout(
                stable: parts.stable,
                tail: parts.tail,
                tailRenderer: .plainText
            )
        }
        if !lastStreamingStable.isEmpty, content.hasPrefix(lastStreamingStable) {
            return Layout(
                stable: lastStreamingStable,
                tail: String(content.dropFirst(lastStreamingStable.count)),
                tailRenderer: .markdown
            )
        }
        return Layout(stable: "", tail: content, tailRenderer: .markdown)
    }
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/StreamingMarkdownContinuityPolicy.kt
package com.tabtin.mobile.features.conversation

/**
 * 流式才切稳定区 / 尾巴。收束不重新切分全文：
 * 若仍是流式期冻住的前缀，只把尾巴转成 Markdown；对不上再整段 parse。
 * 禁止对收束全文再跑 splitter，避免把表格 / 计划拆成两段。
 */
internal object StreamingMarkdownContinuityPolicy {
    internal enum class TailRenderer {
        PlainText,
        Markdown,
    }

    internal data class Layout(
        val stable: String,
        val tail: String,
        val tailRenderer: TailRenderer,
    ) {
        val hasStable: Boolean get() = stable.isNotEmpty()
        val stableIdentity: String get() = stable
    }

    fun layout(
        content: String,
        isStreaming: Boolean,
        lastStreamingStable: String = "",
    ): Layout {
        if (isStreaming) {
            val split = StreamingMarkdownSplitter.split(content)
            return Layout(
                stable = split.stable,
                tail = split.tail,
                tailRenderer = TailRenderer.PlainText,
            )
        }
        if (lastStreamingStable.isNotEmpty() && content.startsWith(lastStreamingStable)) {
            return Layout(
                stable = lastStreamingStable,
                tail = content.substring(lastStreamingStable.length),
                tailRenderer = TailRenderer.Markdown,
            )
        }
        return Layout(
            stable = "",
            tail = content,
            tailRenderer = TailRenderer.Markdown,
        )
    }
}
